base/shell/explorer/startup.cpp

```diff
@@ -332,3 +384,84 @@ ProcessStartupItems(VOID)
 
     return res ? 0 : 101;
 }
+
+static BOOL IsSelfExplorer(VOID)
+{
+    WCHAR szPath[MAX_PATH];
+
+    GetModuleFileName(NULL, szPath, _countof(szPath));
+    PathRemoveArgsW(szPath);
+    PathRemoveBlanksW(szPath);
+
+    return (StrCmpIW(PathFindFileNameW(szPath), L"explorer.exe") == 0);
+}
+
+BOOL DoFinishStartupItems(VOID)
+{
+    if (s_hStartupMutex)
+    {
+        ReleaseMutex(s_hStartupMutex);
+        CloseHandle(s_hStartupMutex);
+        s_hStartupMutex = NULL;
+    }
+    return TRUE;
+}
+
+BOOL DoStartStartupItems(ITrayWindow *Tray)
+{
+    DWORD dwWait;
+
+    if (!IsSelfExplorer())
+        return FALSE;
+
+    if (!s_hStartupMutex)
+    {
+        // We use a mutex to synchronize the auto startup process
+        s_hStartupMutex = CreateMutexW(NULL, FALSE, L"ExplorerIsShellMutex");
```

Comment:
Who are we synchronizing against?
Does Windows have a mutex like this / does it use the same name?

Reply:
In one moment, multiple Explorers can start up. The mutex can synchronize them to protect the Explorer status. I confirmed th mutex name by using API Monitor.

Reply:
The "Shell Explorer" that keeps Desktop must be zero or only one.

Reply:
Sorry, this comment is misleading. I will fix.

Reply:
I fixed the comment.

Reply:
A side comment for information purposes only: this is also what "Emerge Desktop" windows-replacement shell does too:
https://github.com/arthepsy/emergedesktop/blob/master/Source/Explorer/Applet.cpp